portfolio/src/main/java/com/google/sps/servlets/LoginServlet.java

```diff
@@ -0,0 +1,69 @@
+// Copyright 2019 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package com.google.sps.servlets;
+
+import com.google.appengine.api.users.UserService;
+import com.google.appengine.api.users.UserServiceFactory;
+import com.google.appengine.api.datastore.PreparedQuery;
+import com.google.appengine.api.datastore.Query;
+import com.google.appengine.api.datastore.Query.SortDirection;
+import com.google.appengine.api.datastore.DatastoreService;
+import com.google.appengine.api.datastore.DatastoreServiceFactory;
+import com.google.appengine.api.datastore.Entity;
+import java.io.IOException;
+import java.io.PrintWriter;
+import javax.servlet.annotation.WebServlet;
+import javax.servlet.http.HttpServlet;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+
+/** Servlet that checks if user is logged in to display appropriate content */
+@WebServlet("/login")
+public class LoginServlet extends HttpServlet {
+
+    @Override
+    public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
+        response.setContentType("text/html;");
+        PrintWriter out = response.getWriter();
+
+        UserService credentials = UserServiceFactory.getUserService();
+        if(credentials.isUserLoggedIn()) {
+            outputComments(out, credentials);
+        } else {
+            String loginUrl = credentials.createLoginURL("/homepage.html");
+            out.println("<p>Please login <a href=\"" + loginUrl + "\">here.</a></p>");
+        }
+    }
+
+    /** Produces the comment section. */
+    private void outputComments(PrintWriter out, UserService credentials) {
+        String logoutUrl = credentials.createLogoutURL("/homepage.html");
+        out.println("<p>Welcome back, " + credentials.getCurrentUser().getEmail() + ".</p>");
```

Comment:
It would be better to have a separate HTML page that only populates the current user from the server. It can be problematic to generate raw HTML like this and can lead to things like script injections and XSS.

Reply:
The walkthroughs generated raw HTML this way and it was on a separate page like you mentioned. I wanted the server to generate this HTML onto the current page, which has HTML already, to hide the comments and reveal it for only those who are logged in. I was wary of using innerHTML, but innerText wouldn't work - it displayed with quotations and printed verbatim. Is there some other way you/Zach recommend hiding the comment section or getting around innerHTML?

Reply:
You would just use JS to show/hide the commend section or the login section based on what information the server provides, and avoid having the server generate HTML like this. Not only is concatenating user-strings like emails into a server-generated page risky, having the server generate an HTML fragment and inserting it into an external page with javascript/innerHTML means that you end up needing to carefully coordinate the HTML between the outer page and what the server generates to get the right result, which would quickly get difficult in a large project. It also means that the output of the server action is page-specific; if you want to show something else using the same information on a different page, you would have to write a new servlet that returns the same information but formatted into different HTML.

For dynamic pages like this it's better to try to write your server to provide information in a "pure" form like JSON, and just have JavaScript build the page based on the data. That way you can use the same data endpoints and display the same data in different ways on different pages, or only use part of the data if that's all you need for a page. You may have heard the term Model-View-Controller (MVC), and this is something like that. You keep the data (model) separate from the view/controller (how the data is presented to the user) so that each individual part of the code is simpler; the data servlet doesn't have to know or care how the data is being presented, the JS just knows what data it get's and can choose to present it however is appropriate.